### DISCUSSION TITLE
Temp change based on gravity instead of time

### DISCUSSION THREAD

**Discussion by vanosg** [thorrak/fermentrack#630] [Ideas]
This could (maybe) be a cool feature, I haven't given a lot of thought, BUT- what if, when building a fermentation profile, you had the option to use gravity points instead of a set time? That would allow you to start to raise the temperature when it hits 80% of its expected attenuation, for example. So in psuedo code, "hold 66 degrees while gravity >1.020, then raise to 70 degrees over 4 days", as an example. Or even more bonus points, add on "hold 70 degrees until gravity changes <.01 in the last 36 hours, then drop to 45".

I imagine this would be a logical nightmare, but figured it might be worth atleast a discussion as to feasibility. That's one thing that "bugs" me, is I have my awesome and perfect temperature profile all mapped out, and the damn thing finishes fermenting 2 days earlier than I expected and I have to tweak the ferm profile (or just set a constant temp) to do a diacetyl rest. I mean, first world problems, but it would be cool to think about :)

**Comment by thorrak** [comment]
Something I've always wanted to implement are different profile types. I've always imagined three profile types:

**Dumb Profiles** - Basically what exists today. Sit at X temp for Y hours, then move to Z.

**"Smart" Profiles (Temporary profiles?)** - A single-use profile that targets a temperature and a time, and ramps to it. Tracked as #93 

**"Gravity Driven" Profiles** - Profiles that are driven by gravity readings coming off a Tilt/iSpindel. The profile becomes a logical order and series of rules that allow progression to the next step. Some of the logical conditions would be as you suggest - either a % of expected attenuation, a detection of fermentation "stage" (lag, growth, stationary?), or something similar. This is being tracked as #287

**Comment by thorrak** [comment]
Honestly, putting my "solutionizing" hat on, I don't think that gravity driven profiles would be that difficult to implement, as the profile still consists of a series of steps, it's just that each step would have multiple logical conditions that could trigger the next step rather than one (elapsed time).

In my head, you create three objects, all of which have a one-to-many relationship to each other: 
- Gravity Driven Profile
- Step
- Condition

The step would specify the hold temperature, minimum ramp time to hit the hold temperature, and whether the attached conditions are AND or OR. The conditions would specify things such as you recommended (Gravity has dropped 80%, At least 20 days have elapsed).

This would require some additional state tracking during a ferment, but that should be the easy part...
